src/components/coreSelect/index.tsx

```diff
@@ -0,0 +1,215 @@
+}
+
+type Action<TValue> =
+  | { type: 'toggleItem'; payload: TValue }
+  | { type: 'toggleAll' }
+  | { type: 'selectAll' }
+  | { type: 'setItems'; payload: TValue[] };
+
+const reducer =
+  <TValue,>(getItemId: (item: TValue) => string, selectOne: boolean, disallowEmpty: boolean) =>
+  (state: State<TValue>, action: Action<TValue>): State<TValue> => {
+    switch (action.type) {
+      case 'toggleItem': {
+        const itemId = getItemId(action.payload);
+
+        const foundIndex = state.selectedItems.findIndex((x) => x === itemId);
+        let newItems = [...state.selectedItems];
+
+        if (selectOne || state.selectedItems.length === state.items.length) {
+          newItems = [itemId];
+        } else if (foundIndex > -1) {
+          newItems.splice(foundIndex, 1);
+        } else {
+          newItems = [...newItems, itemId].sort((a, b) => (a > b ? 1 : -1));
+        }
+
+        if (disallowEmpty && !newItems.length) {
+          newItems = state.items.map(getItemId);
+        }
+        return { ...state, selectedItems: newItems, allSelected: newItems.length === state.items.length };
+      }
+      case 'toggleAll': {
+        if (!state.allSelected) {
+          return {
+            ...state,
+            selectedItems: state.items.map(getItemId),
+            prevSelectedItems: state.selectedItems,
+            allSelected: true,
+          };
+        }
+        return {
+          ...state,
+          selectedItems: state.prevSelectedItems,
+          allSelected: false,
+        };
+      }
+      case 'selectAll': {
+        return {
+          ...state,
+          selectedItems: state.items.map(getItemId),
+          allSelected: true,
+        };
+      }
+      case 'setItems': {
+        return {
+          ...state,
+          items: action.payload,
+        };
+      }
+      default:
+        return state;
+    }
+  };
+
+/**
+ * Hook that provides selector logic to components. Can be used in controlled or uncontrolled mode. 
+ * In controlled mode this hook keeps its own state of the selection, and in uncontrolled it doesn't.
+ * Uncontrolled mode is activated by passing the prop `getItemSelected` which is then used in determining if an item is selected.
+ * In uncontrolled mode the props `onToggle` and `onToggleAll` can be used to set up event handlers for those actions.
+ */
+const useCoreSelect = <TValue,>({
+  title,
+  items,
+  preselectedItems,
+  getItemTitle,
+  getItemIcon,
+  getItemId,
+  getItemSelected,
+  onToggle,
+  onToggleAll,
+  onChange,
+  disallowEmpty,
+  selectOne,
+}: SelectProps<TValue>): RenderParams<TValue> => {
+  const _getItemTitle = (item: TValue): { title: string; subtitle?: string } => {
+    const _title = getItemTitle(item);
+    return typeof _title === 'string' ? { title: _title, subtitle: undefined } : _title;
+  };
+  const _getItemId = (item: TValue): string => {
+    if (getItemId !== undefined) return getItemId(item);
+    return _getItemTitle(item).title;
```

Comment:
Does this return title since if id is undefined --> title is used to differentiate objects? This could propably return undefined or null, and then do the fallback where function is iniitally called:

```
itemId = _getItemId(..)
if(itemId)
 //do stuff with itemId
else
 itemTitle = _getItemTitle
  //use itemTitle as you would use id
```
would this make sense?

Reply:
I don't see why this needs to be changed. Now if `getItemId` prop is given, it is used to differentiate the items. If it's not, `_getItemTitle` (and inside it the prop `getItemTitle`) is used, which are always defined.

Reply:
I think functions should do what they are namely defined to do, e.q: return the id of the object. If later the same function is used again in the way the name suggests, it may lead that title is used in place of id when it's not suitable anymore.